From 183c4786e570f9038fe8e5b16ebb3b55b8902098 Mon Sep 17 00:00:00 2001
From: liyuting <liyuting2071@phytium.com.cn>
Date: Mon, 20 May 2024 10:03:49 +0800
Subject: [PATCH 057/360] irqchip/irq-gic-v3-its: Enable nonboot CPUs LPIS when
 resume from S3

When system resume from suspend to ram, firmware just restore
GICR_CTLR_ENABLE_LPIs of boot CPU, the other CPUs also should
be restored.

Signed-off-by: Li Yuting <liyuting2071@phytium.com.cn>
Signed-off-by: Wang Xu <wangxu@phytium.com.cn>
Signed-off-by: Feng Jun <fengjun@phytium.com.cn>
Signed-off-by: Chen Baozi <chenbaozi@phytium.com.cn>
Signed-off-by: Wang Yinfeng <wangyinfeng@phytium.com.cn>
---
 drivers/irqchip/irq-gic-v3-its.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/drivers/irqchip/irq-gic-v3-its.c b/drivers/irqchip/irq-gic-v3-its.c
index 0d3a889b1905..6c12a8be046a 100644
--- a/drivers/irqchip/irq-gic-v3-its.c
+++ b/drivers/irqchip/irq-gic-v3-its.c
@@ -4923,6 +4923,7 @@ static void its_restore_enable(void)
 {
 	struct its_node *its;
 	int ret;
+	int cpu;
 
 	raw_spin_lock(&its_lock);
 	list_for_each_entry(its, &its_nodes, entry) {
@@ -4976,6 +4977,23 @@ static void its_restore_enable(void)
 		    GITS_TYPER_HCC(gic_read_typer(base + GITS_TYPER)))
 			its_cpu_init_collection(its);
 	}
+
+	/*
+	 * Enable LPIs:firmware just restore GICR_CTLR_ENABLE_LPIs of boot
+	 * CPU, the other CPUs also should be restored.
+	 */
+	for_each_possible_cpu(cpu) {
+		void __iomem *rbase = gic_data_rdist_cpu(cpu)->rd_base;
+		u32 val;
+
+		/* Enable LPIs */
+		val = readl_relaxed(rbase + GICR_CTLR);
+		if (val & GICR_CTLR_ENABLE_LPIS)
+			continue;
+
+		val |= GICR_CTLR_ENABLE_LPIS;
+		writel_relaxed(val, rbase + GICR_CTLR);
+	}
 	raw_spin_unlock(&its_lock);
 }
 
-- 
2.20.1

